src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs

```diff
@@ -117,17 +122,39 @@ private async Task SerializeToStreamAsyncCore(Stream targetStream, bool async, C
             {
                 if (async)
                 {
-                    await JsonSerializer.SerializeAsync(targetStream, Value, ObjectType, _jsonSerializerOptions, cancellationToken).ConfigureAwait(false);
+                    await JsonSerializer.SerializeAsync(targetStream, Value, _typeInfo, cancellationToken).ConfigureAwait(false);
                 }
                 else
                 {
 #if NETCOREAPP
-                    JsonSerializer.Serialize(targetStream, Value, ObjectType, _jsonSerializerOptions);
+                    JsonSerializer.Serialize(targetStream, Value, _typeInfo);
 #else
-                    Debug.Fail("Synchronous serialization is only supported since .NET 5.0");
+                    Debug.Fail("HttpContent synchronous serialization is only supported since .NET 5.0");
 #endif
                 }
             }
         }
+
+        [RequiresUnreferencedCode(HttpContentJsonExtensions.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(HttpContentJsonExtensions.SerializationDynamicCodeMessage)]
+        private static JsonTypeInfo GetJsonTypeInfo(Type inputType, JsonSerializerOptions? options)
+        {
+            Debug.Assert(inputType is not null);
+
+            // Ensure the options supports the call to GetTypeInfo
+            options ??= JsonHelpers.s_defaultSerializerOptions;
+            options.TypeInfoResolver ??= JsonSerializerOptions.Default.TypeInfoResolver;
+            options.MakeReadOnly();
+
+            return options.GetTypeInfo(inputType);
+        }
+
+        private static void EnsureTypeCompatibility(object? inputValue, Type inputType)
+        {
+            if (inputValue is not null && !inputType.IsAssignableFrom(inputValue.GetType()))
```

Comment:
I realize that this predates your changes, but shouldn't this also be throwing if `inputValue` is null and `inputType` is a value type?

Reply:
Based on a quick code review, the current behavior in STJ is not to throw up-front in `JsonSerializer.ValidateInputType`, but it does throw during serialization in `UnboxOnWrite<T>` with a `JsonException` instead of an `ArgumentException`.

Reply:
>  it could be considered a breaking change

Fair, we probably shouldn't try to address this here.